ui/src/app/pages/workspace/workspace-edit.tsx

```diff
@@ -482,65 +485,83 @@ export const WorkspaceEdit = fp.flow(
-          // entry into billingAccounts so the dropdown entry is not empty.
-          //
-          // The server will not perform an updateBillingInfo call if the received billingAccountName
-          // is the same as what is currently stored.
-          //
-          // This can happen if a workspace is shared to another researcher as an owner.
+        try {
+          const fetchedBillingInfo = await getBillingAccountInfo(
+            this.props.workspace.googleProject
+          );
+          if (
+            !displayBillingAccounts.find(
+              (billingAccount) =>
+                billingAccount.name === fetchedBillingInfo.billingAccountName
+            )
+          ) {
+            // If the user has owner access on the workspace but does not have access to the billing account
+            // that it is attached to, keep the server's current value for billingAccountName and add a shim
+            // entry into billingAccounts so the dropdown entry is not empty.
+            //
+            // The server will not perform an updateBillingInfo call if the received billingAccountName
+            // is the same as what is currently stored.
+            //
+            // This can happen if a workspace is shared to another researcher as an owner.
+            displayBillingAccounts.push({
+              name: this.props.workspace.billingAccountName,
+              displayName: 'User Provided Billing Account',
+              isFreeTier: false,
+              isOpen: true,
+            });
+
+            if (
+              fetchedBillingInfo.billingAccountName !==
+              this.props.workspace.billingAccountName
+            ) {
+              // This should never happen but it means the database is out of sync with Google
+              // and does not have the correct billing account stored.
+              // We cannot send over the correct billing account info since the current user
+              // does not have permissions to set it.
+
+              reportError({
+                name: 'Out of date billing account name',
+                message:
+                  `Workspace ${this.props.workspace.namespace} has an out of date billing account name. ` +
+                  `Stored value is ${this.props.workspace.billingAccountName}. ` +
+                  `True value is ${fetchedBillingInfo.billingAccountName}`,
+              });
+            }
+            this.setState({ billingAccountFetched: true });
+          } else {
+            // Otherwise, use this as an opportunity to sync the fetched billing account name from
+            // the source of truth, Google
+            this.setState((prevState) =>
+              fp.set(
+                ['workspace', 'billingAccountName'],
+                fetchedBillingInfo.billingAccountName,
+                prevState
+              )
+            );
+          }
+          this.setState({ billingAccountFetched: true });
+        } catch (e) {
```

Comment:
it can be hard to follow when a try-block is so long.  Would it make sense to extract it to a new function?

Reply:
I gave it a shot. The logic in this try block was updating an array and  updating state. It seemed like more than I wanted to have in a single function. Please let me know if you think this can be improved.